File: backend/SharedLib/DTOs/Journal/JournalEntryWithTasksAndFeedbackDto.cs
using SharedLib.DTOs.Task;

namespace SharedLib.DTOs.Journal
{
    public record JournalEntryWithTasksAndFeedbackDto(Guid Id, Guid UserId, DateTime CreatedAt, string Title, string Content, bool IsDeleted, 
        JournalFeedbackDto? feedback, string? feedbackManager, IReadOnlyList<TaskItemDto> LinkedTasks)
    {
        public string IdSnippet => Id.ToString()[..6];
        public string IsDeletedStr => IsDeleted ? "Deleted" : "Active";
        public string CreatedAtStr => CreatedAt.ToShortDateString();
        public bool IsFeedbackGiven => feedback is not null;
        public string IsFeedbackGivenStr => IsFeedbackGiven ? "Yes" : "No";

        public string IsFeedbackSeenByUser => IsFeedbackGiven && feedback is not null ? feedback.SeenByUserStr : "-";

        public string[] LinkedTaskTitles => [.. LinkedTasks.Select(x => x.Title)];
        public string LinkedTaskTitlesCsv => LinkedTaskTitles != null && LinkedTaskTitles.Any() ? string.Join(", ", LinkedTaskTitles) : string.Empty;
        public string ContentSnippet => Content.Length <= 120 ? Content : Content.Substring(0, 120) + "...";        // If journals can be long, sending full content for every row is wasteful.
       
    }
}

File: frontend/devpulse-ui/src/app/core/models/journal-entry-with-tasks-and-feedback.dto.ts
import { BaseDto } from './base.dto';
import { TaskItemDto } from './task-item.dto';
import { UserAccountDto } from './user-account.dto';

export interface JournalEntryWithTasksAndFeedbackDto extends BaseDto {
  //id: string; // Guid coming from BaseDto
  idSnippet: string;
  userId: string; // Guid
  createdAt: string; // ISO date string
  title: string;
  content: string;
  isDeleted: boolean;

  feedback: JournalFeedbackDto | null;
  feedbackManager: string | null;

  linkedTasks: TaskItemDto[];

  // Computed fields (sent from backend)
  isDeletedStr: string;
  createdAtStr: string;
  isFeedbackGiven: boolean;
  isFeedbackGivenStr: string;
  isFeedbackSeenByUser: string;
  linkedTaskTitles: string[];
  linkedTaskTitlesCsv: string;
  contentSnippet: string;
}

export interface TeamJournalEntryWithTasksAndFeedbackDto extends JournalEntryWithTasksAndFeedbackDto {
  user: UserAccountDto;

  // Computed fields (sent from backend)
  userDisplayName: string;
}

export interface JournalFeedbackDto extends BaseDto {
  //id: string; // Guid coming from BaseDto
  journalEntryId: string;
  feedbackManagerId: string;
  comment: string;
  createdAt: string;
  seenByUser: boolean;

  // Computed
  createdAtStr: string;
  seenByUserStr: string;
}

export interface CreateJournalDto {
  addJournalEntryDto: AddJournalEntryDto;
  linkedTaskIds: string[]; // HashSet<Guid> → string[]
}

export interface AddJournalEntryDto {
  userId: string; // Guid → string
  title: string;
  content: string;
}
